internal/webhook/external/receiver.go

```diff
@@ -0,0 +1,182 @@
+package external
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+// WebhookReceiver is an interface for components that handle inbound webhooks.
+type WebhookReceiver interface {
+	// getReceiverType returns the type of this receiver.
+	getReceiverType() string
+	// getSecretName returns the name of the Secret upon which this receiver
+	// relies.
+	getSecretName() string
+	// getSecretValues extracts a list of receiver-specific values from the
+	// provided Secret data.
+	getSecretValues(map[string][]byte) ([]string, error)
+	// setSecretData sets the Secret data for this receiver. This is used to
+	// later when handling inbound webhooks.
+	setSecretData(map[string][]byte)
+	// setDetails sets the details of the WebhookReceiver in the form of
+	// kargoapi.WebhookReceiverDetails.
+	setDetails(kargoapi.WebhookReceiverDetails)
+	// GetDetails returns the details of the WebhookReceiver in the form of
+	// kargoapi.WebhookReceiverDetails.
+	GetDetails() kargoapi.WebhookReceiverDetails
+	// GetHandler returns an http.HandlerFunc for handling inbound webhook
+	// requests.
+	GetHandler() http.HandlerFunc
+}
+
+// baseWebhookReceiver is a base implementation of WebhookReceiver that provides
+// common functionality for all WebhookReceiver implementations. It is not
+// intended to be used directly.
+type baseWebhookReceiver struct {
+	client     client.Client
+	project    string
+	secretName string
+	secretData map[string][]byte
+	details    kargoapi.WebhookReceiverDetails
+}
+
+// getSecretName implements WebhookReceiver.
+func (b *baseWebhookReceiver) getSecretName() string {
+	return b.secretName
+}
+
+// setSecretData implements WebhookReceiver.
+func (b *baseWebhookReceiver) setSecretData(secretData map[string][]byte) {
+	b.secretData = secretData
+}
+
+// setDetails implements WebhookReceiver.
+func (b *baseWebhookReceiver) setDetails(
+	details kargoapi.WebhookReceiverDetails,
+) {
+	b.details = details
+}
+
+// GetDetails implements WebhookReceiver.
+func (b *baseWebhookReceiver) GetDetails() kargoapi.WebhookReceiverDetails {
+	return b.details
+}
+
+// NewReceiver returns an appropriate implementation of WebhookReceiver based on
+// the provided kargoapi.WebhookReceiverConfig.
+func NewReceiver(
+	ctx context.Context,
+	c client.Client,
+	baseURL string,
+	project string,
+	secretsNamespace string,
+	cfg kargoapi.WebhookReceiverConfig,
+) (WebhookReceiver, error) {
+	// Pick an appropriate WebhookReceiver implementation based on the
+	// configuration provided.
+	var receiver WebhookReceiver
+	switch {
+	case cfg.GitHub != nil:
+		receiver = newGitHubWebhookReceiver(c, project, cfg)
+	default:
+		return nil, errors.New(
+			"WebhookReceiverConfig has no configuration for a known receiver type",
+		)
+	}
```

Comment:
When the number of implementations grows, I think we would benefit from a registry/factory pattern. It might be worth taking note of this here.